CORDA-4104 Implement custom serialization scheme discovery mechanism

williamvigorr3 opened the pull request:
Discovers a CustomSerializationScheme implementation inside the drivers dir using a system property. Currently only one such scheme can be registered at any one time. This can be controlled by the system property experimental.corda.customSerializationScheme. This must be set to the fully qualified name of the class which implements `CustomSerializationScheme`.

williamvigorr3 committed:
CORDA-4104 Implement CustomSerializationScheme Discovery Mechanism

Discovers a single CustomSerializationScheme implementation inside
the drivers dir using a system property.

williamvigorr3 committed:
Started MockNetwork test

williamvigorr3 committed:
Add driver test of Custom Serialization Scheme

JamesHR3 commented on core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt:
```diff
@@ -43,6 +52,7 @@ interface SerializationEnvironment {
     val checkpointContext: CheckpointSerializationContext
 }
 
+
```
Might as well remove this whitespace change.

williamvigorr3 replied:
Done 🤦

JamesHR3 commented on core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt:
```diff
@@ -31,6 +33,13 @@ interface SerializationEnvironment {
+
+        const val SERIALIZATION_SCHEME_ID_SIZE = 4
+
+        fun getCustomSerializationMagicFromSchemeId(schemeId: Int) : SerializationMagic {
```
Not sure this is the best place for this, as it's not really part of the serialization environment and more of a serialization utility. There's some stuff in `SerializationHelpers.kt` so maybe it could go in there? That also stops us pulling serialization internal classes into other packages.

williamvigorr3 replied:
I agree with that (I was unsure where to put this). Unfortunately it cannot go in `SerializationHelpers.kt` because that is the `Serialization` project. It needs to go somewhere inside 'Core', so it can be used by both `TransactionBuilder.toWireTransaction` in core and `CustomSerializationSchemeAdapter` in `node-api`. We can either leave it here or put it in another file on its own (the package net.corda.core.serialization.internal).

JamesHR3 replied:
My mistake. I think a new file in `net.corda.core.serialization` makes sense for this (I'd have thought `internal` was things not meant to be imported elsewhere).

williamvigorr3 replied:
Hmm I thought `internal` is for things which we don't want to put in the public API for CorDApps (like this). I see your point though.

JamesHR3 commented on core-tests/src/test/kotlin/net/corda/coretests/transactions/TransactionBuilderTest.kt:
```diff
@@ -299,4 +302,21 @@ class TransactionBuilderTest {
+                .addOutputState(outputState)
+                .addCommand(DummyCommandData, notary.owningKey)
+
+        assertFailsWith<UnsupportedOperationException> {
```
Do you know what this exception says when you get it? Is it obvious that the problem is that the corresponding serialization scheme isn't registered? I'm wondering if we can throw a more descriptive exception here.

williamvigorr3 replied:
`java.lang.UnsupportedOperationException: Serialization scheme ([43555300000007], P2P) not supported.` Which is probably quite confusing. We can search in the blob for 0x435553 which is "CUS" in UTF-8 and throw something nicer.

JamesHR3 commented on node/src/main/kotlin/net/corda/node/internal/Node.kt:
```diff
@@ -662,6 +670,21 @@ open class Node(configuration: NodeConfiguration,
+        if (!schemaClass.interfaces.contains(CustomSerializationScheme::class.java)) {
+            throw ConfigurationException("$className does not implement ${CustomSerializationScheme::class.java.canonicalName}")
+        }
+        if (schemaClass.constructors.size != 1 || schemaClass.constructors.single().parameters.isNotEmpty()) {
```
Is this condition overly restrictive? I suppose we can only use a no arg constructor anyway, but maybe it's ok scan for one, rather than enforce there's only one constructor with no arguments. (Happy to hear arguments why this is a bad idea.)

williamvigorr3 replied:
I don't think this matters too much. I am not sure that how writing a constructor which takes arguments would be useful in practise (when you must have a no arg one) but I might be wrong. I have changed it to scan for a no arg one instead of require that there is only one.

JamesHR3 commented on node/src/main/kotlin/net/corda/node/internal/Node.kt:
```diff
@@ -662,6 +670,21 @@ open class Node(configuration: NodeConfiguration,
+            throw ConfigurationException("Custom serialization scheme $className could not be found.")
+        }
+        if (!schemaClass.interfaces.contains(CustomSerializationScheme::class.java)) {
+            throw ConfigurationException("$className does not implement ${CustomSerializationScheme::class.java.canonicalName}")
```
We should probably mention that this has come from the custom serialization scheme feature (so something like `"$className was declared as a custom serialization scheme but does not implement ${CustomSerializationScheme::class.java.canonicalName}"`). Could also potentially inherit from `ConfigurationException` to throw a `CustomSerializationSchemeConfigurationException` or something, but that might be overkill.

williamvigorr3 replied:
Done

JamesHR3 reviewed (changes requested):
Couple of (minor) thoughts.

williamvigorr3 committed:
Fix detekt and minor style error

williamvigorr3 committed:
Respond to review comments

Allow non-single arg constructors (there must be one no args
constructor), move code from SerializationEnviroment into its
own file, improve exceptions to be more user friendly.

williamvigorr3 committed:
Fix minor bug in Scheme finding code  + improve error messages

JamesHR3 reviewed (approved):
LGTM
